pkg/cable/vxlan/vxlan.go

```diff
@@ -0,0 +1,533 @@
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package vxlan
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/cable"
+	"github.com/submariner-io/submariner/pkg/natdiscovery"
+	"github.com/submariner-io/submariner/pkg/types"
+	"github.com/submariner-io/submariner/pkg/util"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog"
+)
+
+const (
+	VxLANIface             = "vxlan-tunnel"
+	VxLANPort              = 4800
+	VxLANOverhead          = 50
+	VxLANVTepNetworkPrefix = 241
+	CableDriverName        = "vxlan"
+	TableID                = 100
+)
+
+type Operation int
+
+const (
+	Add Operation = iota
+	Delete
+	Flush
+)
+
+type vxLan struct {
+	localEndpoint types.SubmarinerEndpoint
+	connections   []v1.Connection
+	mutex         sync.Mutex
+	vxLanIface    *vxLanIface
+}
+
+type vxLanIface struct {
+	activeEndpointHostname string
+	link                   *netlink.Vxlan
+	vtepIP                 net.IP
+}
+
+type vxLanAttributes struct {
+	name     string
+	vxlanID  int
+	group    net.IP
+	srcAddr  net.IP
+	vtepPort int
+	mtu      int
+}
+
+func init() {
+	cable.AddDriver(CableDriverName, NewDriver)
+}
+
+func NewDriver(localEndpoint types.SubmarinerEndpoint, localCluster types.SubmarinerCluster) (cable.Driver, error) {
+	var err error
+
+	v := vxLan{
+		localEndpoint: localEndpoint,
+	}
+
+	if err = v.createVxLANInterface(localEndpoint.Spec.Hostname); err != nil {
+		return nil, fmt.Errorf("failed to setup Vxlan link: %v", err)
+	}
+
+	return &v, nil
+}
+
+func (v *vxLan) createVxLANInterface(activeEndPoint string) error {
+	ipAddr := v.localEndpoint.Spec.PrivateIP
+
+	vtepIP, err := v.getVxlanVtepIPAddress(ipAddr)
+	if err != nil {
+		return fmt.Errorf("failed to derive the vxlan vtepIP for %s, %v", ipAddr, err)
+	}
+
+	defaultHostIface, err := util.GetDefaultGatewayInterface()
+
+	if err != nil {
+		return errors.Wrapf(err, "Unable to find the default interface on host: %s",
+			v.localEndpoint.Spec.Hostname)
+	}
+
+	// Derive the MTU based on the default outgoing interface
+	vxlanMtu := defaultHostIface.MTU - VxLANOverhead
+
+	attrs := &vxLanAttributes{
+		name:     VxLANIface,
+		vxlanID:  1000,
+		group:    nil,
+		srcAddr:  nil,
+		vtepPort: VxLANPort,
```

Comment:
Please look at how wireguard or libreswan handle the UDP port to be used, as it's a configurable parameter.

We should use the UDP encapsulation port configured for this gateway.

Reply:
Done.